vernac/canonical.ml

```diff
@@ -11,29 +11,32 @@ open Names
 open Libobject
 open Recordops
 
-let open_canonical_structure i (_, o) =
+let open_canonical_structure i (_, (o,_)) =
   let env = Global.env () in
   let sigma = Evd.from_env env in
   if Int.equal i 1 then register_canonical_structure env sigma ~warn:false o
 
-let cache_canonical_structure (_, o) =
+let cache_canonical_structure (_, (o,_)) =
   let env = Global.env () in
   let sigma = Evd.from_env env in
   register_canonical_structure ~warn:true env sigma o
 
-let discharge_canonical_structure (_,x) = Some x
+let discharge_canonical_structure (_,(x, local)) =
+  match local with
+  | None | Some false -> Some (x, local)
+  | Some true -> None
 
-let inCanonStruc : Constant.t * inductive -> obj =
+let inCanonStruc : (Constant.t * inductive) * bool option -> obj =
   declare_object {(default_object "CANONICAL-STRUCTURE") with
                   open_function = open_canonical_structure;
                   cache_function = cache_canonical_structure;
-                  subst_function = (fun (subst,c) -> subst_canonical_structure subst c);
+                  subst_function = (fun (subst,(c,locality)) -> subst_canonical_structure subst c, locality);
                   classify_function = (fun x -> Substitute x);
                   discharge_function = discharge_canonical_structure }
 
 let add_canonical_structure x = Lib.add_anonymous_leaf (inCanonStruc x)
 
-let declare_canonical_structure ref =
+let declare_canonical_structure ?local ref =
```

Comment:
Shouldn't this be `?(local=false)`?

Reply:
No, `Some false` is not `None`. According to @maximedenes we have 3 states:
- not given
- given `Local`
- given `Global`

Main point being that one may want to warn/error if something *given* is in conflict with the system behavior (eg `Global` on a section local variable, see the other PR), but be silent if it was the default.

I did not implement these checks, but if go back to 2 states you *can't* implement them.

Reply:
If you do `?(local=false)` you can implement them later without changing the public API, and until then you don't confuse the reader by having multiple equivalent values in your object. 

(Indeed even after you implement the checks they should probably be before the object, and the object should only see the checked locality and doesn't care if it was implicit or explicit.)

Reply:
I think you are right, I pushed the 3 value too deep.